CLN: Standardize and move list_tables method to backend

datapythonista opened the pull request:
xref #2678

Unifying the signatures of `list_tables`, reuse code as much as possible, and move from `Client` to `Backend`.

jreback commented:
can you rebase

jreback commented on ibis/backends/base/client.py:
```diff
@@ -136,22 +139,5 @@ def table(self, name: str) -> ir.TableExpr:
-
-    def _qualify_like(self, like: str | None) -> str | None:
-        return like
+    def list_tables(self, like=None):
```
why does this have a different signature?

datapythonista replied:
This is `Database.list_tables()`. I didn't add the `database` argument here, since it feels a bit strange to do `con.database('foo').list_tables(database='bar')`. I can add it if you want to have the same signature, but in any case, `Database` is set for deprecation, so no big deal.

datapythonista replied:
This operation is in a `Database` object.

What we have in this PR now is:
```python
>>> import ibis
>>> con = ibis.sqlite.connect(...)
>>> con.list_tables()  # `database` is the current database
>>> con.list_tables(database='bar')  # I can specify a different database when listing tables from the connection/client
>>> con.database('bar').list_tables()  # another way to list tables from a different database, equivalent to the previous line
>>> con.database('bar').list_tables(database='foobar')  # DOES NOT WORK. This could work if we make this signature the same as the rest.
```

What does need to be resolved? Do you want to make the last line work? I'm happy to add a `database` argument here if you want, but personally I don't think offering this to the user adds value.

jreback replied:
i actually don't care about the last line, instead about a consistent signature. I don't think the database arg adds *any* value. so either remove it everywhere or have it everywhere.

datapythonista replied:
Ok, I'll add it here for now, and will create an issue to discuss if there are better ways to support databases different than the default. Thanks!

jreback commented on ibis/backends/base/sql/alchemy/client.py:
```diff
@@ -262,35 +262,8 @@ def list_schemas(self):
-        if like is not None:
-            names = [x for x in names if like in x]
-        return sorted(names)
+    def list_tables(self, like=None, database=None, schema=None):
```
this signature is completely different that the others

datapythonista replied:
This is to keep code backward-compatible, since in SQLAlchemy backends use `schema='foo'` instead of `database='foo'`. I see that to pass the test I removed the deprecation code of the `schema` argument and I haven't restored it. If you don't think it's a big deal, I'll just remove `schema`, much simpler.

jreback reviewed (changes requested):
these methods are having completely different signatures. is there a reason?

pep8speaks commented:
Hello @datapythonista! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:









* In the file [`ibis/backends/base/sql/alchemy/client.py`](https://github.com/ibis-project/ibis/blob/7e2ef085944b168dcdaf35535bc5e1ebb7ab78ba/ibis/backends/base/sql/alchemy/client.py):

> [Line 10:1](https://github.com/ibis-project/ibis/blob/7e2ef085944b168dcdaf35535bc5e1ebb7ab78ba/ibis/backends/base/sql/alchemy/client.py#L10): [F401](https://duckduckgo.com/?q=pep8%20F401) 'ibis.expr.operations as ops' imported but unused
> [Line 14:1](https://github.com/ibis-project/ibis/blob/7e2ef085944b168dcdaf35535bc5e1ebb7ab78ba/ibis/backends/base/sql/alchemy/client.py#L14): [F401](https://duckduckgo.com/?q=pep8%20F401) 'ibis.backends.base.Database' imported but unused

datapythonista commented:
@jreback this should be ready now. I removed differences in the signature of SQLAlchemy backend, which didn't make sense (kept them before for backward-compatibility). And cleaned some of the related schema stuff (postgres supports databases and schemas, and most of it was removed from the base SQLAlchemy code, but few things were left).

There are two signatures that are still different:
- File backends (would be nice to remove the path parameters, which is cumbersome, as discussed with Phillip, but better in a follow up, after further discussion)
- `Database.list_tables`, where passing again a `database` parameter is redundant. And it's being removed, so probably not worth anyway. Can add if you have a strong opinion about it, but I'd leave it as is.

Let me know if this looks good now. Thanks!

jreback commented on ibis/backends/base/file/__init__.py:
```diff
@@ -134,6 +134,25 @@ def connect(self, path):
         return pd.__version__
 
+    def list_tables(self, path=None, like=None):
+        """
```
if you override the doc-string then you need to make a full doc-string. also pls type all args

datapythonista replied:
Good point, updated.

jreback reviewed (approved):
lgtm. if there are actually any user facing changes can you list them in a release note @datapythonista
